source/app/app-service-container.ts

```diff
@@ -0,0 +1,182 @@
+/**
+ * @ignore
+ * BEGIN HEADER
+ *
+ * Contains:        AppServiceContainer class
+ * CVM-Role:        Controller
+ * Maintainer:      Hendrik Erz
+ * License:         GNU GPL v3
+ *
+ * Description:     The application service container holds references to
+ *                  singleton instances of every service provider.
+ *
+ * END HEADER
+ */
+
+// Providers
+import AppearanceProvider from '@providers/appearance'
+import AssetsProvider from '@providers/assets'
+import CiteprocProvider from '@providers/citeproc'
+import CommandProvider from '@providers/commands'
+import ConfigProvider from '@providers/config'
+import CssProvider from '@providers/css'
+import DictionaryProvider from '@providers/dictionary'
+import DocumentManager from '@providers/documents'
+import FSAL from '@providers/fsal'
+import LinkProvider from '@providers/links'
+import LogProvider from '@providers/log'
+import MenuProvider from '@providers/menu'
+import NotificationProvider from '@providers/notifications'
+import ProviderContract from '@providers/provider-contract'
+import RecentDocumentsProvider from '@providers/recent-docs'
+import StatsProvider from '@providers/stats'
+import TagProvider from '@providers/tags'
+import TargetProvider from '@providers/targets'
+import TranslationProvider from '@providers/translations'
+import TrayProvider from '@providers/tray'
+import UpdateProvider from '@providers/updates'
+import WindowProvider from '@providers/windows'
+
+export default class AppServiceContainer {
+  private readonly _appearanceProvider: AppearanceProvider
+  private readonly _assetsProvider: AssetsProvider
+  private readonly _citeprocProvider: CiteprocProvider
+  private readonly _commandProvider: CommandProvider
+  private readonly _configProvider: ConfigProvider
+  private readonly _cssProvider: CssProvider
+  private readonly _dictionaryProvider: DictionaryProvider
+  private readonly _linkProvider: LinkProvider
+  private readonly _logProvider: LogProvider
+  private readonly _menuProvider: MenuProvider
+  private readonly _notificationProvider: NotificationProvider
+  private readonly _recentDocsProvider: RecentDocumentsProvider
+  private readonly _statsProvider: StatsProvider
+  private readonly _tagProvider: TagProvider
+  private readonly _targetProvider: TargetProvider
+  private readonly _translationProvider: TranslationProvider
+  private readonly _trayProvider: TrayProvider
+  private readonly _updateProvider: UpdateProvider
+  private readonly _windowProvider: WindowProvider
+  private readonly _fsal: FSAL
+  private readonly _documentManager: DocumentManager
+
+  constructor () {
+    // NOTE: The order in which the providers are instantiated still matters!
```

Comment:
Why is this?

Reply:
Because there are still some race conditions, but I'll have an additional look now that I have the knowledge that the PR is mostly good to think about those race conditions.

Reply:
I've now completely removed any async code from the constructors. I still think we should instantiate the log provider and config provider first, since they can be queried first by other providers, but the rest should not really matter anymore.